main/src/ca/uwaterloo/flix/language/ast/Kind.scala

```diff
@@ -38,7 +38,12 @@ sealed trait Kind {
 object Kind {
 
   /**
-    * The kind of all nullary type expressions.
+    * The kind of type variables whose kinds are not yet determined.
+    */
+  case object Unbound extends Kind
```

Comment:
I think this is the crux of the matter. Ultimately, I don't think we will be able to get by with just "Unbound". Instead, I think we will have to introduce a kind variable and track them in the typer.

For example, you could imagine a type signature:

```
def f(a -> e & e)
```

where e is used both as a proper type and as an effect. Maybe it can't happen so "easily", but I could imagine complicated types where we assign "Unbound" to "e" but end up using it with different kinds.

/That said/,

I think this is a very good starting point and could be refactored towards the right solution eventually (e.g. by replacing `Unbound` by `Var(x: Int)` and so forth.